### ISSUE TITLE
[opentitanlib] add JTAG IO capability to CW310 transport

### ISSUE THREAD

**Issue by timothytrippel** [lowRISC/opentitan#17490]
**For _Background_ and _Problem_ statements see #17489.**

# Solution

A more ergonomic solution to enable writing host-side opentitanlib test programs that can drive LC transitions via the JTAG interface is to add a JTAG IO capability to the opentitanlib CW310 transport.

# Requirements

The JTAG IO interface should:

Addressed in #17614:
- [x] spawn OpenOCD server process that will connect to [Olimex JTAG dongle](https://www.olimex.com/Products/ARM/JTAG/_resources/ARM-USB-TINY_and_TINY_H_manual.pdf) USB endpoint
- [x] open a socket/connect to a PTY to communicate with the OpenOCD server
- [x] send commands directly to OpenOCD server
- [x] shutdown the openocd server upon program termination

Addressed in TBD:
- [x] support the following initial set of [OpenOCD commands](https://openocd.org/doc/html/General-Commands.html#General-Commands):
    - [x] version
    - [x] set_reg
    - [x] get_reg
    - [x] write_memory
    - [x] read_memory
    - [x] halt
    - [x] resume
    - [x] shutdown
- [ ] support the following set of [OpenOCD commands](https://openocd.org/doc/html/General-Commands.html#General-Commands) as a stretch goal:
    - [ ] bp
    - [ ] rpb
- [x] support higher level OT-specific commands:
    - [x] perform LC transition
    - [x] assert a register equals a specified value
    - [x] poll until a register equals a specified value

**Comment by timothytrippel** [comment]
Essentially we want to integrate some of the actions in this test #17029 into opentitanlib.

**Comment by timothytrippel** [comment]
# Tasks (WIP)

- [x] add JTAG IO trait and params to opentitanlib that defines desired (#17590)  
- [ ] implement Jtag trait for CW310 transport in opentitanlib
- [ ] implement Jtag trait for HyperDebug transport in opentitanlib

**Comment by jwnrt** [comment]
> Tasks (WIP)

 I think https://github.com/lowRISC/opentitan/pull/17614 will resolve `#2` and `#3`, with some extra TODOs:

- [ ] Handle and test JTAG to the RISC-V TAP.
- [ ] Check the right TAP is selected when running certain commands (LC TAP only supports `{read,write}_lc_ctr_reg` and not `{read,write}_memory`, `halt`, or `resume`.
- [ ] Detect OpenOCD server startup more robustly by polling for some signal over stdout.

**Comment by timothytrippel** [comment]
> > Tasks (WIP)
> 
> I think #17614 will resolve `#2` and `#3`, with some extra TODOs:
> 
> * [x]  Handle and test JTAG to the RISC-V TAP.
> * [ ]  Check the right TAP is selected when running certain commands (LC TAP only supports `{read,write}_lc_ctr_reg` and not `{read,write}_memory`, `halt`, or `resume`.

@jwnrt is the above task the only one that is not done? Once it is done, can we close this issue?

> * [x]  Detect OpenOCD server startup more robustly by polling for some signal over stdout.
> * [x]  Handle OpenOCD's stdout (currently OpenOCD will also print its socket responses to stdout which can get messy).

**Comment by jwnrt** [comment]
> Since the last task was super easy, I implemented it in #17866 so we can close this.

Oh great, thanks! Sorry for never getting around to it myself
